Deploy-CAM.ps1

```diff
@@ -2924,58 +2924,101 @@ else {
+                $domainNameMessage = "`nPlease enter new fully qualified domain name of the domain which will be created, including a '.' such as example.com"
             }
             else {
-                $domainNameMessage = "Please enter the fully qualified domain name of the domain to connect to. The name must include a '.' such as example.com"
+                $domainNameMessage = "`nPlease enter the fully qualified domain name of the domain to connect to. The name must include a '.' such as example.com"
             }
             Write-Host $domainNameMessage
             $domainName = Read-Host "Domain name"
         }
 
-        # https://social.technet.microsoft.com/Forums/scriptcenter/en-US/db2d8388-f2c2-4f67-9f84-c17b060504e1/regex-for-computer-fqdn?forum=winserverpowershell
-        if (-not $($domainName -imatch '(?=^.{1,254}$)(^(?:(?!\d+\.|-)[a-zA-Z0-9_\-]{1,63}(?<!-)\.?)+(?:[a-zA-Z]{2,})$)')) {
-            Write-Host "The domain name must include two or more components separated by a '.'"
+        if (-not $($domainName -match '^[a-zA-Z0-9]{1,61}[a-zA-Z0-9](?:\.[a-zA-Z]{2,})+$')) {
+            Write-Host "`nInvalid Domain name. Please see https://support.microsoft.com/en-ca/help/909264/naming-conventions-in-active-directory-for-computers-domains-sites-and for valid domain names." -ForegroundColor Red -BackgroundColor Yellow
             $domainName = $null
         }
+
     } while (-not $domainName)
 
+    # Username
+    $username = $null
+    $firstRun = 1
     do {
-        if ( -not $domainAdminCredential ) {
+        if ($domainAdminCredential -and $firstRun) {
+            $username = $domainAdminCredential.UserName
+            $firstRun = $null
+        }
+
+        if ( -not $username) {
             if( -not $deployOverDC ) {
                 $domainAdminMessage = "Please enter the new domain administrator username for the new domain being created"
             }
             else {
                 $domainAdminMessage = "Please enter the username of the service account for Cloud Access Manager to use with the connected domain"
             }
+            $username = Read-Host $domainAdminMessage
+        }
 
-            $domainAdminCredential = Get-Credential -Message $domainAdminMessage
-            $confirmedPassword = Read-Host -AsSecureString "Please re-enter the password"
-
-            # Need plaintext password to check if same
-            $clearPassword = ConvertTo-Plaintext $confirmedPassword
-            if (-not ($domainAdminCredential.GetNetworkCredential().Password -ceq $clearPassword)) {
-                # don't match- try again.
-                Write-Host "The entered passwords do not match."
-                $domainAdminCredential = $null
+        # only check if it is not deployOverDC
+        if (-not $deployOverDC) {
+            if ((-not ($username -match '^[A-Za-z0-9]+(?:[_-][A-Za-z0-9]+)*$')) -or ($username.Length -gt 20)) {
```

Comment:
I think some of the same comments apply to this regex:

- use `imatch` and remove uppercase character ranges
- should use `\d` for digits

Reply:
And just for interest- how is this different than imatch '\w+'?

Reply:
@bill-dall Good catch. The rules for usernames are less restrictive so I think `\w+` would do fine. :)

Reply:
\w+ does not check for -

Reply:
Also, it seems you can't start or end with - or _

Reply:
- should not match with \w. IS - accepted in usernames?

Reply:
I'm fine with whichever regex works here. As long as it's tested.